### PR TITLE
Daffy-New-Deal Perception + Control Merged, LF working

### PR DESCRIPTION
**Pull request by AleksandarPetrov**
## Description

Merged and working packages for lane following. Lane following works. Some of these packages still need cleanup, conversion to DTROS, and documentation. Some debug nodes were disabled to make it work.

### The following packages are (somewhat cleaned up and) functional:
- anti_instagram
- ground_projection
- lane_control
- lane_filter
- line_detector
- stop_line_filter
- vehicle_detection
- led_detector
- led_detector

### Combined the following PRs/branches
- https://github.com/duckietown/dt-core/pull/56 (https://github.com/duckietown/dt-core/tree/daffy-edit-groundproj)
- https://github.com/duckietown/dt-core/pull/55 (https://github.com/duckietown/dt-core/tree/daffy-edit-controller)
- https://github.com/duckietown/dt-core/pull/54 (https://github.com/duckietown/dt-core/tree/daffy-edit-lane-filter-and-more)
- https://github.com/duckietown/dt-core/pull/48 (https://github.com/duckietown/dt-core/tree/daffy-edit-anti-instagram)
- https://github.com/duckietown/dt-core/tree/daffy-VehDet-StopLine
- https://github.com/duckietown/dt-core/tree/daffy-edit-leds-merge

Lane following demo works! ^..^
Stopping at stop line /vehicle detection doesn't work though, @gibernas can you take a look?

## Type of change

Please mark the relevant options with X:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ X Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change is constrained only to one package, any interfaces to nodes from other packages are maintained
- [ ] This change is constrained only to packages in this repository, any interfaces to nodes from other repositories are maintained
- [X] This change required a documentation update
- [ ] This change is only a documentation update
- [ ] This change restructures the repository (at level higher than package level)
- [ ] This change requires a change in the Duckiebook (if yes, please describe and, if possible, make a pull request for an edit in the book and link it here)

## Tests

Please describe all tests ran and their outcome. It is highly recommended that you at least do the tests listed by default, even if you only changed the documentation.

- [X] Made the Duckiebot move with keyboard control
- [X] Visually checked the published camera stream
- [X] Verified that the lane following demo works
- [ ] Verified that the indefinite navigation demo works

## Style compliance
Please make sure that all your changes conform to the Duckietown (code style)[link] and code (documentation style)[link] guidelines:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I checked that this pull request is compliant with the Duckietown code style
- [ ] I checked that this pull request is compliant with the Duckietown code documentation style

## Next steps
Please describe if you have any questions, comments, or doubts towards the Duckietown software guardians regarding implementation, code style, running tests, documentation, etc.

## FOR THE REVIEWERS
__IMPORTANT:__ DO NOT CHANGE THIS SECTION!

- [ ] Functionality (is the change necessary and is it implemented in the most appropriate way): @liampaull or @afdaniele
- [ ] Integration testing (verify the tests and run additional if necessary): @?
- [ ] Code style compliance: @gibernas
- [ ] Documentation compliance: @aleksandarpetrov

_To the reviewers:_ If you deem necessary, request code edits, more tests, documentation update, or additional reviewing. __Do not review your own code!__

_PR Template, last edit 23 Aug 2019_

### CODE REVIEW

**Comment by liampaull** [comment]
@AleksandarPetrov is this a replacement for #48 #54 #39 ? if so let's close those other PRs without merging ASAP so we don't get confused.

**Comment by AleksandarPetrov** [comment]
@liampaull Done ;)

**Comment by AleksandarPetrov** [comment]
@gibernas, must have been a mistake, I put them back now. Thanks for the catch :)

**Comment by liampaull** [comment]
@AleksandarPetrov can you tell me what branch all of the other repositories should be in to test this?

**Comment by AleksandarPetrov** [comment]
@liampaull try with `daffy-new-deal` of both the duckiebot and car interfaces.

**Comment by liampaull** [comment]
I get this error:
```
[ERROR] [1592327534.850407]: bad callback: <bound method DTSubscriber._monitored_callback of <duckietown.dtros.dtsubscriber.DTSubscriber object at 0x7f9e56018cd0>>
Traceback (most recent call last):
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/topics.py", line 750, in _invoke_callback
    cb(msg)
  File "/duckietown/catkin_ws/src/dt-ros-commons/packages/duckietown/include/duckietown/dtros/dtsubscriber.py", line 125, in _monitored_callback
    out = self._user_callback(*args, **kwargs)
  File "/duckietown/catkin_ws/src/dt-core/packages/ground_projection/src/ground_projection_node.py", line 128, in lineseglist_cb
    new_segment.points[0] = self.pixel_msg_to_ground_msg(received_segment.pixels_normalized[0])
  File "/duckietown/catkin_ws/src/dt-core/packages/ground_projection/src/ground_projection_node.py", line 99, in pixel_msg_to_ground_msg
    pixel = self.ground_projector.vector2pixel(norm_pt)
AttributeError: 'NoneType' object has no attribute 'vector2pixel'```

**Comment by AleksandarPetrov** [comment]
@liampaull I guess you were unlucky in a race condition. Should be fixed now, can you try again?

**Comment by liampaull** [comment]
ok the error is gone. The performance in the simulator is very poor. It doesn't look like the lane pose localization is working at all. Also, all of the debugging symbols and outputs have changed, are they described anywhere?

File: packages/ground_projection/src/ground_projection_node.py [hunk 1]
```diff
@@ -1,90 +1,287 @@
...
+        node_name (:obj:`str`): a unique, descriptive name for the node that ROS will use
 
+    Subscribers:
+        ~camera_info (:obj:`sensor_msgs.msg.CameraInfo`): Intrinsic properties of the camera. Needed for rectifying the segments.
+        ~lineseglist_in (:obj:`duckietown_msgs.msg.SegmentList`): Line segments in pixel space from unrectified images
 
-class GroundProjectionNode(object):
+    Publishers:
+        ~lineseglist_out (:obj:`duckietown_msgs.msg.SegmentList`): Line segments in the ground plane relative to the robot origin
+        ~debug/ground_projection_image/compressed (:obj:`sensor_msgs.msg.CompressedImage`): Debug image that shows the robot relative to the projected segments. Useful to check if the extrinsic calibration is accurate.
```
Comment (liampaull) [inline packages/ground_projection/src/ground_projection_node.py:34]: I am never able to see anything on this output.

Comment (liampaull) [reply packages/ground_projection/src/ground_projection_node.py:34]: I think it's caused by this error:
```
[ERROR] [1592339452.245835]: bad callback: <bound method DTSubscriber._monitored_callback of <duckietown.dtros.dtsubscriber.DTSubscriber object at 0x7f1344374950>>
Traceback (most recent call last):
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/topics.py", line 750, in _invoke_callback
    cb(msg)
  File "/duckietown/catkin_ws/src/dt-ros-commons/packages/duckietown/include/duckietown/dtros/dtsubscriber.py", line 125, in _monitored_callback
    out = self._user_callback(*args, **kwargs)
  File "/duckietown/catkin_ws/src/dt-core/packages/ground_projection/src/ground_projection_node.py", line 140, in lineseglist_cb
    debug_image_msg = self.bridge.cv2_to_compressed_imgmsg(self.debug_image(seglist_out))
  File "/duckietown/catkin_ws/src/dt-core/packages/ground_projection/src/ground_projection_node.py", line 276, in debug_image
    pt1=(int(segment.points[0].y * -400) + 200, int(segment.points[0].x * -400) + 300),
ValueError: cannot convert float NaN to integer
```

**Comment by liampaull** [comment]
here's what the lane filter belief image looks like now: 
image
this is completely wrong

File: packages/ground_projection/src/ground_projection_node.py (hunk 1 above)
Comment (liampaull) [reply packages/ground_projection/src/ground_projection_node.py:34]: This was actually caused by an error in the the way the simulator was linked to ROS. It was not publishing valid camera_info's and I think that the camera_info is now being used more explicitly inside the ground_projection. img_width and height were 0 causing 'nan's. We could test for this (kind of a good idea whenever you are dividing by something)

**Comment by liampaull** [comment]
seems the forward velocity in the simulator is much less

**Comment by liampaull** [comment]
I had to tune my robot gain to 1.5 to make it overcome friction and actually move.

**Comment by liampaull** [comment]
the performance in my home lab is pretty poor. It has always been pretty poor though. I will try again tomorrow in better lighting conditions and if it seems nominal I will merge.

**Pull request merged**